pytential/linalg/proxy.py

```diff
@@ -68,15 +71,23 @@ def partition_by_nodes(
     :arg max_particles_in_box: passed to :class:`boxtree.TreeBuilder`.
     """
 
+    if dofdesc is None:
+        dofdesc = places.auto_source
+    dofdesc = sym.as_dofdesc(dofdesc)
+
     if max_particles_in_box is None:
-        # FIXME: this is just an arbitrary value
-        max_particles_in_box = 32
+        max_particles_in_box = _DEFAULT_MAX_PARTICLES_IN_BOX
 
-    if tree_kind is not None:
-        from boxtree import box_flags_enum
-        from boxtree import TreeBuilder
+    lpot_source = places.get_geometry(dofdesc.geometry)
+    discr = places.get_discretization(dofdesc.geometry, dofdesc.discr_stage)
 
-        builder = TreeBuilder(actx.context)
+    if tree_kind is not None:
+        from pytential.qbx import QBXLayerPotentialSource
+        if isinstance(lpot_source, QBXLayerPotentialSource):
+            builder = lpot_source.tree_code_container.build_tree()
+        else:
+            from boxtree import TreeBuilder
+            builder = TreeBuilder(actx.context)
```

Comment:
Explain or remove re-create-y case.

Reply:
@inducer Wanted to do this, but wasn't sure which array context: the one the user has or the `_setup_actx` of the `QBXLayerPotentialSource`?

Reply:
In the meantime, the latest rebase removed the `if` since nobody was using the fallback.

This is mostly good to go on my end and we can move the several code containers that are now attached to `QBXLayerPotentialSource` in a separate PR?

Reply:
#145